src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,184 @@
+
+/*
+`?rankup` calculates the number of katas of each rank required for a user to rank up.
+This command is just for fun, however should increase competition between users,
+and (hopefully) increase motivation.
+
+If no username is passed, server nickname of user is taken instead.
+
+Options:
+  <username>   The username to check. Defaults to users discord server nickname
+  --target=    The rank to reach, or a username of a user to overtake. Defaults to users next rank
+  --language=  The language to inspect. Defaults to "overall"
+  --mode=      Calculation mode.
+                  Least favours 1kyus for least overall katas.
+                  Each finds number of katas per rank individually
+                  Spread (default) prioritises high katas, but also distributes points to lower ones
+  --limit=     The max rank to use in calculations ([1-8]kyu)
+*/
+
+enum Mode {
+  Least = "least",
+  Each = "each",
+  Spread = "spread", // default
+}
+const DEFAULTMODE: Mode = Mode.Spread;
+const SPREADWEIGHT: number = 0.6; // used for Spread mode
+
+// Convert synonyms into the names used by CW API
+const LANGSYNONYMS: { [lang: string]: string } = {
+  py: "python",
+  js: "javascript",
+  hs: "haskell",
+  brainfuck: "bf",
+  cs: "coffeescript",
+  "c#": "csharp",
+  ts: "typescript",
+  "c++": "cpp",
+  "f#": "fsharp",
+  objectivec: "objc",
+  "obj-c": "objc",
+  "objective-c": "objc",
+  coldfusionmarkuplanguage: "cfml",
+  coldfusion: "cfml",
+};
+
+// https://docs.codewars.com/gamification/ranks
+// 3-8 dan added speculatively, based on trend
+const RANKTHRESHOLDS: { [rank: string]: number } = {
+  "7kyu": 20,
+  "6kyu": 76,
+  "5kyu": 229,
+  "4kyu": 643,
+  "3kyu": 1768,
+  "2kyu": 4829,
+  "1kyu": 13147,
+  "1dan": 35759,
+  "2dan": 97225,
+  "3dan": 264324,
+  "4dan": 718593,
+  "5dan": 1953551,
+  "6dan": 5310860,
+  "7dan": 14437910,
+  "8dan": 39250354,
+};
+
+const RANKPOINTS: { [rank: string]: number } = {
+  "8kyu": 2,
+  "7kyu": 3,
+  "6kyu": 8,
+  "5kyu": 21,
+  "4kyu": 55,
+  "3kyu": 149,
+  "2kyu": 404,
+  "1kyu": 1097,
+};
+
+const capitalise = (s: string): string =>
+  s.length ? s[0].toUpperCase() + s.slice(1).toLowerCase() : "";
+
+const getLang = (lang: string): string => LANGSYNONYMS[lang.toLowerCase()] || lang.toLowerCase();
+
+// Retrieve a users language score, or overall score if language is undefined
+const getUserLangScore = async (user: string, lang?: string): Promise<number | undefined> => {
+  return await fetch("https://www.codewars.com/api/v1/users/" + user)
+    .then((response: any): any => (response.status === 404 ? { success: false } : response.json()))
+    .then((result: any): any => {
+      if (result.success === false) return;
+      return (lang ? result.ranks.languages[getLang(lang)] : result.ranks.overall)?.score ?? 0;
+    });
+};
+
+export default async (message: Message, args: CommandArg[], opts: Record<string, string>) => {
```

Comment:
You can add `Options` type (as long as values are string), and change the type of `opts`:

```typescript
export type Options = {
  // note that these are optional with `?`
  mode?: string;
  language?: string;
  target?: string;
  limit?: string;
};

export default async (message: Message, args: CommandArg[], opts: Options) => {
```

This is better because it forces you to handle the undefined cases correctly.

For example, the code in this PR is not affected, but `RegExp.test` stringifies the argument, so
```javascript
const opts = {};
/un/.test(opts.foo); //=> true
```